tools/hsmtools.c

```diff
@@ -0,0 +1,214 @@
+#include <bitcoin/privkey.h>
+#include <ccan/str/str.h>
+#include <ccan/err/err.h>
+#include <ccan/noerr/noerr.h>
+#include <ccan/read_write_all/read_write_all.h>
+#include <common/utils.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sodium.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#define ERROR_HSM_FILE errno
+#define ERROR_LIBSODIUM 3
+
+static void show_usage(void)
+{
+	printf("./hsmtools <method> [arguments]\n");
+	printf("methods:\n");
+	printf("	- decrypthsm <path/to/hsm_secret> <password>\n");
+	exit(0);
+}
+
+static bool ensure_hsm_secret_exists(int fd)
+{
+	if (fsync(fd) != 0) {
+		close(fd);
+		return false;
+	}
+	if (close(fd) != 0)
+		return false;
+
+	fd = open(".", O_RDONLY);
+	if (fd < 0)
+		return false;
+	if (fsync(fd) != 0) {
+		close(fd);
+		return false;
+	}
+
+	close(fd);
+	return true;
+}
+
+static int decrypt_hsm(const char *hsm_secret_path, const char *passwd)
+{
+	int fd;
+	struct stat st;
+	struct secret key, hsm_secret;
+	u8 salt[16] = "c-lightning\0\0\0\0\0";
+	crypto_secretstream_xchacha20poly1305_state crypto_state;
+	u8 header[crypto_secretstream_xchacha20poly1305_HEADERBYTES];
+	/* The cipher size is static with xchacha20poly1305. */
+	u8 cipher[sizeof(struct secret) + crypto_secretstream_xchacha20poly1305_ABYTES];
+
+	/* We can use a relative path as libplugin chdir into lightning_dir
+	 * before calling us. */
+	if (stat(hsm_secret_path, &st) != 0)
+		errx(ERROR_HSM_FILE, "Could not stat hsm_secret");
+	if (st.st_size <= 32)
+		errx(ERROR_HSM_FILE, "hsm_secret is not encrypted");
+	fd = open(hsm_secret_path, O_RDONLY);
+	if (fd < 0)
+		errx(ERROR_HSM_FILE, "Could not open hsm_secret");
+
+	if (!read_all(fd, header, crypto_secretstream_xchacha20poly1305_HEADERBYTES))
+		errx(ERROR_HSM_FILE, "Could not read cipher header");
+	if (!read_all(fd, cipher, sizeof(cipher)))
+		errx(ERROR_HSM_FILE, "Could not read cipher body");
+
+	/* Derive the encryption key from the password provided, and try to decrypt
+	 * the cipher. */
+	if (crypto_pwhash(key.data, sizeof(key.data), passwd, strlen(passwd), salt,
+	                  crypto_pwhash_argon2id_OPSLIMIT_MODERATE,
+	                  crypto_pwhash_argon2id_MEMLIMIT_MODERATE,
+	                  crypto_pwhash_ALG_ARGON2ID13) != 0)
+		errx(ERROR_LIBSODIUM, "Could not derive a key from the password.");
+	if (crypto_secretstream_xchacha20poly1305_init_pull(&crypto_state, header,
+	                                                    key.data) != 0)
+		errx(ERROR_LIBSODIUM, "Could not initialize the crypto state");
+	if (crypto_secretstream_xchacha20poly1305_pull(&crypto_state, hsm_secret.data,
+	                                               NULL, 0, cipher, sizeof(cipher),
+	                                               NULL, 0) != 0)
+		errx(ERROR_LIBSODIUM, "Could not retrieve the seed. Wrong password ?");
+	close(fd);
+
+	/* Create a backup file, "just in case". */
+	rename(hsm_secret_path, "hsm_secret.backup");
```

Comment:
does `rename` update just the file name, or the entire path?

Reply:
entire path

Reply:
a bit of a stretch goal, but it'd be nice if it's in the same directory as the original hsm file, so that it's easy to find/recover in case of a problem

Reply:
Actually it was in plugin version (we chdir in libplugin.c at `init`) but I kept it although we don"t run it from the same directory.
In case of a not-already handled problem, the backup is still in the running directory.. And it would be quite hacky, just for a tool  : recovering the path to the lightning-dir from the hsm_secret path, then prefixing the rename..